File: leetcode/303-range-sum-query-immutable/javascript/num-array.js
let NumArray = function(nums) {
    this.nums = [nums[0]];
    for (let i = 1; i < nums.length; i += 1) {
        this.nums[i] = this.nums[i - 1] + nums[i];
    }
};


NumArray.prototype.sumRange = function(left, right) {
    if (left === 0) {
        return this.nums[right];
    }
    return this.nums[right] - this.nums[left - 1];
};

let numArray = new NumArray([-2, 0, 3, -5, 2, -1]);
console.log(numArray.sumRange(0, 2));
console.log(numArray.sumRange(2, 5));
console.log(numArray.sumRange(0, 5));

File: leetcode/303-range-sum-query-immutable/python/num_array.py
class NumArray:
    
    def __init__(self, nums: list[int]):
        total = 0
        for i in range(len(nums)):
            nums[i] += total
            total = nums[i]
        self.nums = nums
    
    
    def sumRange(self, left: int, right: int) -> int:
        if left == 0:
            return self.nums[right]
        return self.nums[right] - self.nums[left - 1]


if __name__ == '__main__':
    numArr = NumArray([-2, 0, 3, -5, 2, -1])
    print(numArr.sumRange(0, 2))
    print(numArr.sumRange(2, 5))
    print(numArr.sumRange(0, 5))
